From d7cbbe0664ac370c415f1981b633216f8bffcbec Mon Sep 17 00:00:00 2001
From: jan janasek <jjanasek@redhat.com>
Date: Thu, 25 Jun 2026 12:21:20 +0200
Subject: [PATCH] policy: add container_runtime_nnp_domtrans interface

Systemd services using standard hardening directives (such as
DynamicUser, ProtectSystem, or RestrictSUIDSGID) automatically
have NoNewPrivileges (NNP) enabled by systemd.

When these services (e.g., Podman Quadlets) attempt to start
containers, the kernel blocks the SELinux domain transition to
container_runtime_t due to NNP or nosuid mount restrictions,
causing the containers to fail at startup.

This commit introduces the `container_runtime_nnp_domtrans`
interface. It allows caller domains to explicitly bypass NNP
and nosuid restrictions on the process2 class during the
transition to the container runtime. This provides a clean
way for base policies to enable secure, systemd-managed
container workloads.

Signed-off-by: Jan Janasek <jjanasek@redhat.com>
---
 container.if | 20 ++++++++++++++++++++
 1 file changed, 20 insertions(+)

diff --git a/container.if b/container.if
index b6499af..516945e 100644
--- a/container.if
+++ b/container.if
@@ -1127,3 +1127,23 @@ interface(`container_signull',`
 
         allow $1 container_t:process signull;
 ')
+
+########################################
+## <summary>
+##	Execute container_runtime in the container_runtime domain, 
+##	and allow the transition under NoNewPrivileges (NNP) or nosuid.
+## </summary>
+## <param name="domain">
+##	<summary>
+##	Domain allowed to transition.
+##	</summary>
+## </param>
+#
+interface(`container_runtime_nnp_domtrans',`
+	gen_require(`
+		type container_runtime_t;
+	')
+
+	container_runtime_domtrans($1)
+	allow $1 container_runtime_t:process2 { nnp_transition nosuid_transition };
+')
\ No newline at end of file
